File: llamaratti/llamaratti/DragDropTextView.m
/**
 * @file DragDropTextView.m
 *
 * @brief NSTextView subclass for drag/drop and other custom operations
 *
 * @author Created by Geoff G. on 05/25/2025
 *
 */

#import "DragDropTextView.h"
#import "ViewController.h"

#import "NSColor+More.h"
#import "NSImage+More.h"

#import "Shared.h"
#import "Utils.h"

@implementation DragDropTextView
{
    BOOL _dragAborted;
    
    NSColor *_savedBorderColor;
    NSInteger _savedBorderWidth;
}

/**
 * @brief Initializes user interface elements
 *
 */
- (void)awakeFromNib
{
    _dragAccepted=NO;
    _dragAborted=NO;
    
    _savedBorderColor=[NSColor colorWithCGColor:[[self layer] borderColor]];
    _savedBorderWidth=[[self layer] borderWidth];
    
    [self setWantsLayer:YES];

    [self registerForDraggedTypes:@[NSPasteboardTypeFileURL]];
}

/**
 * @brief drawRect
 *
 */
- (void)drawRect:(NSRect)dirtyRect {
    [super drawRect:dirtyRect];
    
}

#pragma mark - NSDraggingDestination

/**
 * @brief prepareForDragOperation
 *
 */
- (BOOL)prepareForDragOperation:(id<NSDraggingInfo>)sender {
    
    return _dragAccepted;
}

/**
 * @brief performDragOperation
 *
 * @returns whether we have accepted the operation
 */
- (BOOL)performDragOperation:(id<NSDraggingInfo>)sender {
    
    return _dragAccepted;
}

/**
 * @brief draggingEntered
 *
 * Highlights the text field when files are dragged into the control
 *
 * @returns NSDragOperation
 */
- (NSDragOperation)draggingEntered:(id<NSDraggingInfo>)sender
{
    // Are we accepting drag operations?
    if ( !_dragAccepted ) {
        return NSDragOperationNone;
    }
    
    // Yes, highlight the control
    [[self layer] setBorderColor:[[NSColor colorNamed:@"HighlightColor"] CGColor]];
    [[self layer] setBorderWidth:2];
    _dragAborted=NO;
    
    return NSDragOperationEvery;
}

/**
 * @brief draggingUpdated
 *
 * @returns NSDragOperation
 */
- (NSDragOperation) draggingUpdated:(id<NSDraggingInfo>)sender
{
    // Are we accepting drag operations?
    if ( !_dragAccepted ) {
        return NSDragOperationNone;
    }
    return NSDragOperationEvery;
}

/**
 * @brief Called when the dragging operation has ended
 *
 */
- (void)draggingExited:(id<NSDraggingInfo>)sender
{
    // Unhighlight the control
    [[self layer] setBorderColor:[_savedBorderColor CGColor]];
    [[self layer] setBorderWidth:_savedBorderWidth];

    _dragAborted=YES;
}

/**
 * @brief Validates dragged files & adds them to the current context
 *
 * Validates the files dragged into the control and adds them to the
 * current context
 *
 * @param sender - NSDraggingInfo
 *
 */
- (void)draggingEnded:(id<NSDraggingInfo>)sender
{
    // Reset our highlighted border
    [[self layer] setBorderColor:[_savedBorderColor CGColor]];
    [[self layer] setBorderWidth:_savedBorderWidth];
    
    // Are we accepting drag operations?
    if ( !_dragAccepted || _dragAborted ) {
        return;
    }
    
    // Do we have our parent view controller?
    if ( !_vcParent || !isInstanceOfClass(_vcParent, [NSViewController class]) ) {
        return;
    }

    // Are we busy?
    ViewController *vc=(ViewController *)_vcParent;
    if ( [vc isBusy] ) {
        return;
    }

    // Do we have a LlamarattiWrapper?
    LlamarattiWrapper *llamaWrapper=[vc llamaWrapper];
    if ( llamaWrapper == nil ) {
        return;
    }

    // Do we have any dragged files?
    NSPasteboard *pb = [sender draggingPasteboard];
    if ( ![[pb types] containsObject:NSPasteboardTypeFileURL] ) {
        return;
    }
    
    // We use these options to access security scoped file URLs
    NSDictionary *options = @{NSPasteboardURLReadingFileURLsOnlyKey:@YES,
                            @"NSPasteboardURLReadingSecurityScopedFileURLsKey":@YES};
    NSArray *arrDraggedURLs = [pb readObjectsForClasses:@[[NSURL class]]
                                                options:options];

    // Did we get a valid array?
    if ( !isValidNSArray(arrDraggedURLs) ) {
        return;
    }
        
    // Add any supported dropped file(s)...
    NSUInteger countAdded=0;
    NSUInteger countMax=[LlamarattiWrapper maxSupportedMedia];
    for ( NSURL *urlTmpDragged in arrDraggedURLs ) {
        
        NSURL *urlDragged = urlTmpDragged;
        
        // Is this a supported audio type?
        if ( [llamaWrapper isSupportedAudioURL:urlDragged] ) {

            if ( [vc loadAudioIntoContext:urlDragged
                         useSecurityScope:YES] ) {
                countAdded++;
            }
        }
        
        // Is this a supported image type?
        else if ( [llamaWrapper isSupportedImageURL:urlDragged] ) {

            BOOL useSecurityScope=YES;
            
            // Is this an HEIC image?
            NSString *pathExt=[urlDragged pathExtension];
            if ( [[pathExt lowercaseString] isEqualToString:@"heic"] ) {
                
                // Yes, can we convert the file?
                urlDragged = [NSImage convertHEICtoTmpJPG:urlTmpDragged
                                            withTmpPrefix:gMediaTemplate
                                                withLoss:IMAGE_CONVERSION_LOSS];
                if ( !urlDragged ) {
                    continue;
                }
                
                // URL points to temp folder within app sandbox
                useSecurityScope=NO;
                
            // Is this a WEBP image?
            } else if ( [[pathExt lowercaseString] isEqualToString:@"webp"] ) {
                
                // Yes, can we convert the file?
                urlDragged = [NSImage convertWEBPtoTmpJPG:urlTmpDragged
                                            withTmpPrefix:gMediaTemplate];
                if ( !urlDragged ) {
                    continue;
                }
                
                // URL points to temp folder within app sandbox
                useSecurityScope=NO;
            }
            
            if ( [vc loadImageIntoContext:urlDragged
                         useSecurityScope:useSecurityScope] ) {
                countAdded++;
            }
        }
        
        // Have we added enough?
        if ( countAdded > countMax ) {
            break;
        }
    }
    
    // Did we add any media files?
    if ( countAdded == 0 ) {
        
        // No, notify...
        NSString *modelName=[llamaWrapper title];
        NSString *suppTypes=[llamaWrapper supportedMediaTypes];
        NSString *title=@"Media Type Not Supported";
        NSString *msg=[NSString stringWithFormat:@"'%@' does not support this media type. Supported types are:\n\n%@\n",modelName,suppTypes];
        [Utils showAlertWithTitle:title
                       andMessage:msg
                         urlTitle:nil
                              url:nil
                          buttons:nil];
    } else {
        
#if RIPPLE_IMAGES_DURING_ADDING
        // Start rippling the images
        [vc startTimedRippleFor:RIPPLE_IMAGES_DURATION];
#endif // RIPPLE_IMAGES_DURING_ADDING
        
        // Hide the drag/drop icon
        [vc toggleDragDropIcon:NO];
    }
}

/**
 * @brief Cleans up the contents of the views and deletes any temporary files
 *
 */

- (void)reset {
    
    NSTextStorage *textStorage = self.textStorage;

    // Remove all embedded subviews
    for (NSView *subview in self.subviews) {
        [subview removeFromSuperview];
    }

    // Remove all attachments
    NSRange fullRange = NSMakeRange(0, textStorage.length);
    [textStorage enumerateAttribute:NSAttachmentAttributeName
                            inRange:fullRange
                            options:0
                         usingBlock:^(id value, NSRange range, BOOL *stop) {
        if ([value isKindOfClass:[NSTextAttachment class]]) {
            [textStorage deleteCharactersInRange:range];
        }
    }];

    // Clear all text and formatting
    [textStorage setAttributedString:[[NSAttributedString alloc] initWithString:@""]];

    // Reset typing attributes (optional)
    self.typingAttributes = @{};
    
    // Remove any temporary files
    NSString *template=[NSString stringWithFormat:@"%@*",gMediaTemplate];
    [Utils removeTempFilesMatchingTemplate:template];
}

/**
 * @brief Handles redisplay when the view is resized
 *
 */
- (void)viewDidEndLiveResize {
    
    [self displayIfNeeded];
}
@end

File: llamaratti-swift/llamaratti-swift/DragDropTextView.swift
/**
 * @file DragDropTextView.swift
 *
 * @brief NSTextView subclass to support drag/drop and other custom stuff
 *
 * @author Created by Geoff G. on 05/25/2025
 *
 */
import Cocoa

class DragDropTextView: NSTextView {

    // Placeholder string is not supported by default in NSTextView
    var placeholderString: String?
    
    var dragAccepted: Bool = false
    var dragAborted: Bool = false
    var vcParent: ViewController?
    
    private var savedBorderColor: CGColor?
    private var savedBorderWidth: CGFloat = 0

    /**
     * @brief Initializes user interface elements
     *
     */
    override func awakeFromNib() {
        super.awakeFromNib()

        wantsLayer = true
        if let layer = self.layer {
            savedBorderColor = layer.borderColor
            savedBorderWidth = layer.borderWidth
        }

        registerForDraggedTypes([.fileURL])
    }

    /**
     * @brief draw
     *
     * Required to update placeholder, which is not supported by NSTextView
     *
     */
    override func draw(_ dirtyRect: NSRect) {
        super.draw(dirtyRect)

        // If there is no text, draw the placeholder
        if string.isEmpty, let placeholder = placeholderString {
            let attributes: [NSAttributedString.Key: Any] = [
                .foregroundColor: NSColor.placeholderTextColor,
                .font: self.font ?? NSFont.systemFont(ofSize: NSFont.systemFontSize)
            ]
            let rect = NSInsetRect(bounds,14, 10)
            placeholder.draw(in: rect,
                withAttributes: attributes)
        }
    }

    /**
     * @brief didChangeText
     *
     * Required to update placeholder, which is not supported by NSTextView
     *
     */
    override func didChangeText() {
        super.didChangeText()
        needsDisplay = true
    }
    
    // MARK: - NSDraggingDestination

    /**
     * @brief prepareForDragOperation
     *
     */
    override func prepareForDragOperation(_ sender: NSDraggingInfo) -> Bool {
        return dragAccepted
    }

    /**
     * @brief performDragOperation
     *
     * @returns whether we have accepted the operation
     */
    override func performDragOperation(_ sender: NSDraggingInfo) -> Bool {
        return dragAccepted
    }

    /**
     * @brief draggingEntered
     *
     * Highlights the text field when files are dragged into the control
     *
     * @returns NSDragOperation
     */
    override func draggingEntered(_ sender: NSDraggingInfo) -> NSDragOperation {
        if ( !dragAccepted ) {
            return []
        }

        layer?.borderColor = NSColor(named: "HighlightColor")?.cgColor
        layer?.borderWidth = 2
        dragAborted = false

        return .every
    }
    
    /**
     * @brief draggingUpdated
     *
     * @returns NSDragOperation
     */
    override func draggingUpdated(_ sender: NSDraggingInfo) -> NSDragOperation {
        return dragAccepted ? .every : []
    }
    
    /**
     * @brief Called when the dragging operation has ended
     *
     */
    override func draggingExited(_ sender: NSDraggingInfo?) {
        layer?.borderColor = savedBorderColor
        layer?.borderWidth = savedBorderWidth
        dragAborted = true
    }

    /**
     * @brief Validates dragged files & adds them to the current context
     *
     * Validates the files dragged into the control and adds them to the
     * current context
     *
     * @param sender - NSDraggingInfo
     *
     */
    override func draggingEnded(_ sender: NSDraggingInfo) {
        
        layer?.borderColor = savedBorderColor
        layer?.borderWidth = savedBorderWidth

        // Are we accepting drag operations?
        if ( !dragAccepted || dragAborted ) {
            return
        }
        
        // Do we have our parent view controller?
        if ( vcParent == nil ) {
            return
        }
        
        // Are we busy?
        if ( vcParent!.isBusy() ) {
            return
        }
        
        // Do we have a Wrapper?
        let llamaWrapper = vcParent?.llamaWrapper
        if ( llamaWrapper == nil ) {
            return
        }
        
        let pb = sender.draggingPasteboard
        guard pb.types?.contains(.fileURL) == true else {
            return
        }

        // Can we get the dragged URLs?
        let options: [NSPasteboard.ReadingOptionKey: Any] = [
            .urlReadingFileURLsOnly: true,
            // Note: The magic option to enable security-scoped drag drop
            NSPasteboard.ReadingOptionKey(rawValue: "NSPasteboardURLReadingSecurityScopedFileURLsKey"): true
        ]
        guard let arrDraggedURLs = pb.readObjects(
                 forClasses: [NSURL.self],
                    options: options) as? [URL], !arrDraggedURLs.isEmpty else {
            return
        }
        
        // Add any supported dropped file(s)...
        var countAdded = 0
        let countMax = LlamarattiWrapper.maxSupportedMedia()
        for url in arrDraggedURLs {
            
            // Is this a supported audio file?
            if ( vcParent!.llamaWrapper!.isSupportedAudioURL(url) ) {
                
                // Can we load the audio into our context?
                if ( vcParent!.loadAudioIntoContext(urlAudio: url) ) {
                    countAdded += 1
                }
            }
            
            // Is this a supported image?
            else if ( llamaWrapper!.isSupportedImageURL(url) ) {
                
                // Can we load the image into our context?
                if ( vcParent!.loadImageIntoContext(urlImage: url) ) {
                    countAdded += 1
                }
            }

            // Have we added enough?
            if countAdded > countMax {
                break
            }
        }

        // Did we add any media files?
        if countAdded == 0 {
            
            // No, notify...
            let modelName = llamaWrapper!.title()
            let suppTypes = vcParent!.llamaWrapper!.supportedMediaTypes()
            let title = "Media Type Not Supported"
            let msg = String(format:"'%@' does not support this media type. Supported types are:\n\n%@\n",modelName!,suppTypes!)
            _ = vcParent!.showAlert(title: title,
                                  message: msg,
                                  buttons: ["Ok"])
        } else {
            // Hide the drag/drop icon
            vcParent?.toggleDragDropIcon(bDisplay: false)
        }
    }
}
